cEP-0020.md

```diff
@@ -0,0 +1,389 @@
+- Badge-IV:
+  - Name: The helper
+  - Details: The one who answers every newcomer questions on Gitter
+  - Activities:
+    - At least 20 good messages to newcomer
+- Badge-V:
+  - Name: The coala Developer
+  - Details: The ones who are in the coala developers team
+  - Activities:
+    - Introduce `.coafile` in other projects
+    - Merge a difficulty/newcomer Pull Request
+    - Review at least a difficulty/newcomer Pull Request
+    - Merge a difficulty/low Pull Request
+    - Review at least a difficulty/low or higher Pull Request
+
+So, for becoming a developer at coala, a newcomer must complete all the
+activities mentioned in the `coala Developer` badge. But in case a newcomer
+is not able to become a developer at coala, he/she could take away the other
+badges.
+
+#### Note
+
+Names and metrics of tracked activities will also be shown on the user
+dashboard along with the number of times they have been done. That will
+help maintainers to check if an automatically awarded badge is correct or
+not. So, in case of doubts on any newcomer, there will be a button to
+revoke those badges.
+
+The `coala Developer` badge doesn't mean that a newcomer is in the coala
+developer team. If a newcomer wants to be the part of the developer team
+after earning the developer badges, he/she will have to ask a maintainer
+on Gitter to verify all the activities they have done, on the dashboard.
+
+## Implementation
+
+### Getting the data
+
+#### GitHub/GitLab
+
+For getting the data related to newcomers activity on GitHub/GitLab to the
+community, we will mainly use [IGitt](https://gitlab.com/gitmate/open-source/IGitt),
+But for getting the data related to review comments on PRs we will use the
+data provided by the [meta-review project](https://github.com/coala/cEPs/blob/e9f7587856e1c236d37e96f54a0cb4d72069776d/cEP-0019.md).
+
+#### Gitter
+
+For getting the stats of newcomer's messages, we will use
+[Gitter API](https://developer.gitter.im/docs/messages-resource) and import
+all the messages to newcomers or by newcomers in our webservices. And then
+we will import those message to our community repos into two Django models
+i.e.: `GoodMessage` and `BadMessage`
+These messages can be analysed and categorized with the use of
+[basic sentiment analysis](http://fjavieralba.com/basic-sentiment-analysis-with-python.html)
+
+### A System for Tracking Activities, Points and Levels
+
+Now that the rules for our gamification design are ready and the relevant data is
+imported in the community repo, a good way to start the implementation of gamification system is to design some models and methods to track the user's activity,
+points and levels:
+
+#### Creating the activity model
+
+An Activity model will give a way to keep a log of each time the user earns
+or loses points so that there will be a record of his points history on a
+dashboard.
+
+A mockup of a basic `Activity` Django model would look something like:
+
+```python
+class Actvity(models.Model):
```

Comment:
What if someone has done multiple activities of the same type? For example, how do you represent "Bob has created **two** newcomer issues"?

Reply:
Nice pick :+1: 
I think I need to have a field named `activities` in the User model which will have a ManyToMany relationship with Activity model(which will also contains a field for `number_of_times_that_activity_has_been_done` and default is set to `0`) and a method named `count_similar_activity` which will be a counter to check if that activity has been done previously, if yes I will increase `number_of_times_that_activity_has_been_done` with one .